### PR TITLE
[VA-API] Fix ffmpeg filters

### PR DESCRIPTION
**Pull request by lbeltrame**
**Changes**
ffmpeg is very picky about the filters to be used when using VA-API,
because most of them are incompatible. This is particularly evident when
burning-in subtitles.

This commit:

- adds new video filters to the VA-API ffmpeg command;
- ensures `hwmap` is also last in the line.

Tests done:

- Played a video file with subtitles
- Played a video file without subtitles

Note that far more testing is required to make sure all corner cases
are doing OK, especially with ffmpeg 4.0 (I only have access to 4.1).

Reference: https://trac.ffmpeg.org/wiki/Hardware/VAAPI

*Issues**
Fixes issue #642.

### CODE REVIEW

**Comment by hawken93** [comment]
Is it possible to get this tested with 4.0? If it breaks we'll have to finally do version checking with ffmpeg to know what command line to pass it based on version...
At this moment it seems more important to me that 4.0 works than 4.1 since 4.1 has other known breakages that got downprioritized in favor of 4.0.

**Comment by lbeltrame** [comment]
I personally can't test with 4.0 as I'm using the ffmpeg installed on the system Jellyfin runs on (openSUSE Tumbleweed) and that is 4.1. Also this requires a working VA-API setup, which means static builds are out.

That said, if anyone has access to a system with 4.0 installed, they can use the same command line as mentioned in the issue, and see if it breaks (that's how I tested my changes before making the PR).

**Comment by lbeltrame** [comment]
2 picoseconds Google search shows that at least those filters exist before 4.0. I have access to a few Ubuntu 18.04 LTS machines but those have too old ffmpeg (3.4.x) to make the comparison.

**Comment by lbeltrame** [comment]
More digging done: the way ffmpeg suggests to use VA-API (see trac wiki link) predates even 4.0, so I'm expecting it to work.

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 1]
```diff
@@ -1743,6 +1743,13 @@ private string GetFixedSizeScalingFilter(Video3DFormat? threedFormat, int reques
 
                 filters.Add(subParam);
 
+		// Ensure proper filters are passed to ffmpeg in case of hardware acceleration via VA-API
```
Comment (JustAMan) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: indentation seems off, you seem to be using tabs instead of spaces

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 2]
```diff
@@ -1743,6 +1743,13 @@ private string GetFixedSizeScalingFilter(Video3DFormat? threedFormat, int reques
...
+		// Reference: https://trac.ffmpeg.org/wiki/Hardware/VAAPI
+		if (string.Equals(outputVideoCodec, "h264_vaapi", StringComparison.OrdinalIgnoreCase))
```
Comment (JustAMan) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: I wonder what happens if a `vp9_vaapi` or `hevc_vaapi` is selected... :)

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs (hunk 1 above)
Comment (lbeltrame) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: This was done hackily on the machine where the serer runs. Will fix.

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs (hunk 2 above)
Comment (lbeltrame) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: To answer this question, I need to tell why I did that in the first place: I looked at the ffmpeg command line that was failing, and then looked at the code *where* those filters were set up. They were set up only with `h264_vaapi` (but I don't know why in the first place).

I can't unfortunately test with vp9 or hevc as the CPU of the server I run Jellyfin on doesn't support them natively.

**Comment by lbeltrame** [comment]
Thanks for the review. I don't know enough about how Jellyfin does things to answer properly the second question (see my comment). Any pointers on where to look?

**Comment by JustAMan** [comment]
> Thanks for the review. I don't know enough about how Jellyfin does things to answer properly the second question (see my comment). Any pointers on where to look?

cc @MatMaul as he did most of the digging for now I think.
As for where to find answers... only code can help for now, we're still in the process of spelunking here.

**Comment by lbeltrame** [comment]
I guess I'll start using `grep` and see where it brings me...

**Comment by lbeltrame** [comment]
@JustAMan Well.. the fun thing is that code I changed is only hit in the case of h264. HEVC and VP9 only set different codec options in ffmpeg, and nothing else.

See MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:1702

**Comment by nvllsvm** [comment]
Temporarily closing - dev branch is being removed in favor of master branch. Also shrinking repo. https://github.com/jellyfin/jellyfin/issues/311

**Comment by lbeltrame** [comment]
Do I need to rebase this?

**Comment by JustAMan** [comment]
@lbeltrame Yes, please rebase against shrunk repo and target `master` branch.

**Comment by lbeltrame** [comment]
Will do that tonight or tomorrow.
